### PR TITLE
Experiment light self contained test runner(no reflection/tcp stack)

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
image

Question PR for Viktor, we'll try to run tests using code generator with xunit attributes and registering test adapter with module initializer, trying to avoid reflection and network TCP at all(only primitive types)

cc:  @ViktorHofer @Evangelink @jkotas

### CODE REVIEW

File: src/testrunner/Microsoft.NET.Test.Runner/Microsoft.NET.Test.Runner.sln [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿
```
Comment (MarcoRossignoli) [inline src/testrunner/Microsoft.NET.Test.Runner/Microsoft.NET.Test.Runner.sln:1]: @ViktorHofer I need to build the test runner package during the build(using private dotnet), can you point me to a good place?

**Comment by jkotas** [comment]
cc @agocke

**Comment by MarcoRossignoli** [comment]
> The source generator that you are proposing to would live in vstest repo and be eventually supported as standard vstest feature. Is that correct?

This can be an idea, @Evangelink will help with that here and can take inspiration by Jeremy one. 
If the experiments will work and the runtime team will decide to opt for this way we'll move to a different repo for sure.

**Comment by agocke** [comment]
Yeah we all seem to be moving in the same direction together in terms of test running and source generators.

I think there are a couple pieces in this area that would be valuable and it might be worth spending some time to get them together.

1. An AOT-compatible generator for XUnit. Note that I've picked out XUnit in particular because it's what we use in the dotnet org. I would expect that the generator would be mostly compatible with the XUnit attributes, although there may be some needed differences. It may (hopefully) be able to re-use the XUnit assert library in its entirety.
2. A fast, lightweight runner for dotnet/runtime tests. This is what we've been building so far. The primary goal, as far as I understand, was to create a runner that supports some of the more unique dotnet/runtime requirements and idioms.
3. Support the models from (1) and (2) in `dotnet test`. As far as I understand this involves two parts. One is to make the top-level VSTest hosting layer allow for execution via a separate exe. The other is to make an "adapter" that can be compiled into a target exe and communicate with the hosting layer. I don't fully understand the existing contract between test frameworks (like xunit) and the vstest adapter, so I don't fully understand what's needed here.

Overall, the biggest open question I see is around (1) and (2). It certainly looks like there's room for these to be combined, although I'm not quite certain of how well the goals of the runtime tests and XUnit are aligned.

The biggest difference between the goals of (1) and (2) that I see is that, ideally, I think (1) is a general project that should be usable by all XUnit test frameworks. One of the main goals I would have in mind of an AOT XUnit runner is that it could be used by the community to provide testing for their apps and libraries in AOT conditions. I think this forces certain design goals more strictly than might be proposed otherwise, especially publication as a separate project and overall compatibility with the XUnit test runner.

**Comment by MarcoRossignoli** [comment]
For the 1 I left to @Evangelink the comprehensive explanation, he did some investigation(I think that we can sync with xunit team also and maybe provide this extension there, it's true that maybe we cannot support all kind of attributes and will be supported only in C#). 
For the 2, we had some experiments with a possible self contained test platform with some specific characteristics:
* no dependencies at all
* no reflection for extensions, move from discovery pattern to registration pattern(with module initializer)
* communication between extensions(in-process, like an xunit adapter and out of process like a datacollector) using json-rpc(plain text), this allow us to avoid coupling, ship a dependency(it's a text protocol implementation) and improve the versioning moving from "binary" extension model(dependency) to a "data message" based model. In-process adapters like xunit, mstest, nunit etc... will communicate using full duplex `Func<string,Task<string>>` in-process callback.
* allow trimming/R2R/self-containing/single file deploy
* allow scenario where there's no runtime/sdk installed

3 With self contained platform there's no more need for `dotnet test` the test project compiles an autonomous exe that can accept command line commands like a normal console(so we can implement built-in trx logger etc..).
It can be used as-is or can be used as "custom host"(feature already available) together with `dotnet test`(vstest.console.exe/dll) but this means some communication between the current runner and the new self host application and today we have tcp that doesn't fit well with runtime repo limit.

We're already experimenting/introducing some of these feature inside the current test platform to improve performance and stability, but I think that we won't be able to remove reflection, dependencies and move to registration pattern, so the idea here is to turn the experiment we had in something that maybe can help these new/future specific scenarios.
NB: at the moment we don't have any plan this is a PoC.

cc: @pavelhorak @nohwnd @jakubch1 @Haplois @cvpoienaru @AbhitejJohn

**Comment by jeffhandley** [comment]
Tagging @dotnet/area-system-text-json for their visibility on this too.

**Comment by MarcoRossignoli** [comment]
cc: @jkoritzinsky

**Pull request closed without merging**